File: Problems/300. Longest Increasing Subsequence/300. Longest Increasing Subsequence.java
import java.util.ArrayList;
import java.util.List;
class Solution {
    private int binarySearch(List<Integer> seq, int target) {
        int l = 0, r = seq.size()-1;

        while (l <= r) {
            int m = l + (r-l)/2;
            int num = seq.get(m);
            if (num == target) return m;
            else if (num < target) l = m + 1;
            else if (num > target) r = m - 1;
        }

        return l;
    }

    public int lengthOfLIS(int[] nums) {
        ArrayList<Integer> subseq = new ArrayList<>();
        for (int num: nums) {
            int pos = binarySearch(subseq, num);
            if (pos == subseq.size()) subseq.add(num);
            else subseq.set(pos, num);
        }

        return subseq.size();
    }
}

File: Problems/300. Longest Increasing Subsequence/300. Longest Increasing Subsequence.py
from typing import List
class Solution:
    def binarySearch(self, seq: List[int], target: int) -> int:
        l, r = 0, len(seq) - 1
        while l <= r:
            m = (r+l) // 2
            if seq[m] == target:
                return m
            elif seq[m] < target:
                l = m + 1
            else:
                r = m - 1
            
        return l
        
    def lengthOfLIS(self, nums: List[int]) -> int:
        layer_min = [nums[0]]
        
        for num in nums:
            pos = self.binarySearch(layer_min, num)
            if pos == len(layer_min): layer_min.append(num)
            else: layer_min[pos] = num
        
        return len(layer_min)